pkg/resolver/multiresolver/multiresolver.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package multiresolver
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"strings"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/resolver"
+	"github.com/ethersphere/bee/pkg/resolver/client/ens"
+	"github.com/ethersphere/bee/pkg/resolver/multiresolver/multierror"
+)
+
+// Ensure MultiResolver implements Resolver interface.
+var _ resolver.Interface = (*MultiResolver)(nil)
+
+var (
+	// ErrTLDTooLong denotes when a TLD in a name exceeds maximum length.
+	ErrTLDTooLong = fmt.Errorf("TLD exceeds maximum length of %d characters", maxTLDLength)
+	// ErrInvalidTLD denotes passing an invalid TLD to the MultiResolver.
+	ErrInvalidTLD = errors.New("invalid TLD")
+	// ErrResolverChainEmpty denotes trying to pop an empty resolver chain.
+	ErrResolverChainEmpty = errors.New("resolver chain empty")
+	// ErrResolverChainFailed denotes that an entire name resolution chain
+	// for a given TLD failed.
+	ErrResolverChainFailed = errors.New("resolver chain failed")
+	// ErrCloseFailed denotes that closing the multiresolver failed.
+	ErrCloseFailed = errors.New("close failed")
+)
+
+type resolverMap map[string][]resolver.Interface
+
+// MultiResolver performs name resolutions based on the TLD label in the name.
+type MultiResolver struct {
+	resolvers resolverMap
+	logger    logging.Logger
+	cfgs      []ConnectionConfig
+	// ForceDefault will force all names to be resolved by the default
+	// resolution chain, regadless of their TLD.
+	ForceDefault bool
+}
+
+// Option is a function that applies an option to a MultiResolver.
+type Option func(*MultiResolver)
+
+// NewMultiResolver will return a new MultiResolver instance.
+func NewMultiResolver(opts ...Option) *MultiResolver {
+	mr := &MultiResolver{
+		resolvers: make(resolverMap),
+	}
+
+	// Apply all options.
+	for _, o := range opts {
+		o(mr)
+	}
+
+	// Discard log output by default.
+	if mr.logger == nil {
+		mr.logger = logging.New(ioutil.Discard, 0)
+	}
+	log := mr.logger
+
+	if len(mr.cfgs) == 0 {
+		log.Info("name resolver: no name resolution service provided")
+		return mr
+	}
+
+	// Attempt to conect to each resolver using the connection string.
+	for _, c := range mr.cfgs {
+
+		// Warn user that the resolver address field is not used.
+		if c.Address != "" {
+			log.Warningf("name resolver: connection string %q contains resolver address field, which is currently unused", c.Address)
+		}
+
+		// Select the appropriate resolver.
+		switch c.TLD {
+		case "eth":
+			mr.connectENSClient(c.TLD, c.Endpoint)
+		case "rsk":
+			mr.connectENSClient(c.TLD, c.Endpoint)
+		case "":
+			mr.connectENSClient("", c.Endpoint)
+		default:
+			log.Errorf("default domain resolution not supported")
+		}
```

Comment:
Do we need this specific tld handling here? Should we just relay what is passed in the config?

Reply:
The case can be useful if different resolver clients are selected based on the TLD. The intent was to use a `RSKClient` for RNS domains etc.

Reply:
I will remove this for now, until it is required.

Reply:
Right, but RNS resolver should be the same as ENS resolver, just needs to be tested. This whole switch statement can be replaced with just `mr.connectENSClient(c.TLD, c.Endpoint)`. We can add more complex handling when is actually needed.